[WIP] Digital Elevation Model (DEM) tutorial (#2985)
gazebosim/gz-sim: tutorials/digital_elevation_models.md

Image reference: https://www.futurelearn.com/info/courses/advanced-archaeological-remote-sensing/0/steps/356833

They are useful for simulation because they offer a simple way to provide realistic digital representation of the terrain.
For fields like robotics, accurate topography can be beneficial for develop algorithms and/or models for autonomous navigation, motion planning, task planning, and control because they can provide cost effective, rapid development cycles.
Engineers can develop and test algorithms in these environments before physical deployment.

In this tutorial we'll dive deeper into what makes heightmaps, where to get them, and how to process them to use for simulation in modern Gazebo.

## Gazebo's supported heightmap types

Modern Gazebo supports image heightmaps, meshes, and Digital Elevation Models (DEMs).
While this tutorial will be focusing more in depth with DEMs, we will provide a quick overview of the other formats available to try out.

### Image heightmaps

Image heightmaps use a 2D grayscale image, where each pixel coresponds to the elevation at that point. Black (or `0`) represents the lowest point and while (or `255`) represents the highest.

> This is a simple and efficient way to store elevation data because less memory compared to meshes and DEMs.

Reviewer: ```suggestion
This is a simple and efficient way to store elevation data because it can use less memory compared to meshes and DEMs.
```

![Cave example](files/digital_elevation_models/mesh_cave.png)

Meshes were the format used for the [Darpa Subterranean Challenge (SubT) simulation](https://github.com/osrf/subt/wiki)
and most were created from point cloud converted data (check out this [tutorial on how to convert point cloud data to a 3D mesh model for Gazebo](http://gazebosim.org/api/sim/10/pointcloud.html)). 
These heightmap meshes can be found [here on fuel](https://app.gazebosim.org/OpenRobotics/fuel/collections/SubT%20Tech%20Repo).

### Digital Elevation Models

A Digital Elevation Model (DEM) is a 3D representation of a terrain's surface that does not include any objects like buildings or vegetation.
DEMs are frequently created by using a combination of sensors, such as satellites, LIDAR, radar, or cameras.
The terrain elevations for ground positions are sampled at regularly-spaced horizontal intervals.

![Monterey bay DEM in Gazebo](files/digital_elevation_models/dem_monterey_bay.png)

<!-- TODO figure out better title -->
# Walkthrough with DEMs for Gazebo

There are two main types of DEM formats, a vector-based Triangular Irregular Network (TIN) or a grid of elevations (raster).

> In Gazebo, maintainers have only tested raster based formats (more specifically GeoTiff or `.tif` files) but in theory Gazebo should be able to run any [GDAL](https://gdal.org/en/stable/index.html) default supported format

Reviewer: minor rephrase:

```suggestion
In Gazebo, raster based formats (more specifically GeoTiff or `.tif` files) are the most tested formats but in theory Gazebo should be able to run any [GDAL](https://gdal.org/en/stable/index.html) default supported format
```

(since GDAL is the backend library that Gazebo uses to read DEM files).

> Let's look at an example of getting DEMs from publically available data and using some common open source tools to them for Gazebo.

Reviewer: Not sure if this is what you intended to say:

```suggestion
Let's look at an example of getting DEMs from publicly available data and using some common open source tools to process them for Gazebo.
```

Using [QGIS](https://qgis.org/),

* (optional) When working with earth DEMs, it can be helpful to add a Google Satellite layer (as the base layer) to verify the imported DEM is in the correct projection.

  * In the top toolbar, go to Layer > Add Layer > Add XYZ Connection...
  * In the dropdown, select Google Satellite > Click 'Add' > then 'Close'

* In the top toolbar, go to Layer > Add Layer > Add Raster Layer...
* Select the file that we downloaded and unzipped (from the instructions above) > Click 'Add' > then 'Close'
* In the bottom bar, update "Scale" to `1:50000` and in the "Coordinate" textbot, enter the lat/lon coordinates retrieved from USGS earlier (i.e., `37.7444°, -119.5341°`) > hit the Enter key (this will move the viewpoint to the desired location)

  * If the optional step was followed, QGIS should look something like the image below, otherwise only the DEM will be displayed

![QGIS with Google Satellite as base layer and DEM](files/digital_elevation_models/qgis_satellite_and_dem.png)

* If the Processing Toolbox panel isn't open, click the gear icon in the top toolbar

  * Search for "Clip raster by extent" and double click to open the dialog

>   * Input layer: should be the imported geotif

Reviewer: capitalized for consistency

```suggestion
  * Input layer: should be the imported GeoTiff
```

This is the `x` and `y` values that need to be entered in the `<size>` tag. To calculate z, subtract the `Minimum` from the `Maximum` elevation (i.e., `z = Maximum - Minimum`).

Optionally, to move the model down to Gazebo's camera default viewpoint, update the `z` of the `<pos>` tag to be the negative of the `Maximum` elevation (i.e., `<pos>0 0 -Maximum</pos>`). Note when using seabed data (e.g., GEBCO), negate the `Minimum` instead.

The model should look similar to the following:

```xml
<model name="half_dome">
  <static>true</static>
  <link name="link">
      <collision name="collision">
          <geometry>
              <heightmap>
                  <uri>half_dome.tif</uri>
                  <!-- gdalinfo -stats <dem_file> -->
                  <size>3793 3563 1481.398</size>
                  <pos>0 0 -2694.987</pos> <!-- -(max or min elevation) -->
              </heightmap>
          </geometry>
      </collision>
      <visual name="visual">
          <geometry>
              <heightmap>
                  <use_terrain_paging>false</use_terrain_paging>
                  <texture>
                      <diffuse>materials/textures/rocks_diffuse.png</diffuse>

>                       <normal>materials/textures/rocks_normal.png</normal>

Reviewer: looks like these are placed in the `textures` directory (no parent `materials` dir) inside `half_dome_example`. So we probably needs to either:
* update these to `textures/rocks_diffuse.png`, or
* add a parent `materials` directory

                      <size>250</size>
                  </texture>
                  <blend>
                      <min_height>2</min_height>
                      <fade_dist>5</fade_dist>
                  </blend>
                  <blend>
                      <min_height>4</min_height>
                      <fade_dist>5</fade_dist>
                  </blend>
                  <uri>half_dome.tif</uri>
                  <!-- gdalinfo -stats <dem_file> -->
                  <size>3793 3563 1481.398</size>
                  <pos>0 0 -2694.987</pos> <!-- -(min or max elevation) -->
              </heightmap>
          </geometry>
      </visual>
  </link>
</model>
```

> Launch Gazebo and it should look something like this:

Reviewer: Add some instructions?

I think the users will need to be in the `half_dome_example` for the textures to load 

```bash
cd tutorials/files/digital_elevation_models/half_dome_example
gz sim -v 4 half_dome.sdf
```
